From 3898ade87ecf3bfe258716745ea8d631787ca6f5 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 21 Aug 2025 09:11:53 +0800
Subject: [PATCH 234/236] mtk: hostapd: properly call ap_sta_remove_link_sta()
 before ap_free_sta()

If the function ap_free_sta() is called only for a single link of a
STA MLD, especially the setup link, it might cause a use-after-free
issue.

The function ap_sta_remove_link_sta should be called before
ap_free_sta() to prevent the issue.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/sta_info.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index e32c491c9..46145cf32 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -524,6 +524,8 @@ void hostapd_free_stas(struct hostapd_data *hapd)
 		sta = sta->next;
 		wpa_printf(MSG_DEBUG, "Removing station " MACSTR,
 			   MAC2STR(prev->addr));
+		if (ap_sta_is_mld(hapd, prev))
+			ap_sta_remove_link_sta(hapd, prev);
 		ap_free_sta(hapd, prev);
 	}
 }
@@ -768,6 +770,8 @@ static void ap_handle_session_timer(void *eloop_ctx, void *timeout_ctx)
 		       "session timeout");
 	sta->acct_terminate_cause =
 		RADIUS_ACCT_TERMINATE_CAUSE_SESSION_TIMEOUT;
+	if (ap_sta_is_mld(hapd, sta))
+		ap_sta_remove_link_sta(hapd, sta);
 	ap_free_sta(hapd, sta);
 }
 
@@ -842,8 +846,11 @@ static void ap_sta_assoc_timeout(void *eloop_ctx, void *timeout_ctx)
 	if (sta->flags & WLAN_STA_AUTH)
 		ap_sta_deauthenticate(hapd, sta,
 				      WLAN_REASON_PREV_AUTH_NOT_VALID);
-	else
+	else {
+		if (ap_sta_is_mld(hapd, sta))
+			ap_sta_remove_link_sta(hapd, sta);
 		ap_free_sta(hapd, sta);
+	}
 }
 
 
-- 
2.45.2

